app/models/container_deployment.rb

```diff
@@ -6,5 +6,314 @@ class ContainerDeployment < ApplicationRecord
+           end
+      ip_array << ip if ip
+    end
+    ip_array
+  end
+
+  def add_deployment_provider(options)
+    provider = ExtManagementSystem.new(
+      :type      => options[:provider_type],
+      :name      => options[:provider_name],
+      :port      => options[:provider_port],
+      :hostname  => options[:provider_hostname],
+      :ipaddress => options[:provider_ipaddress])
+    provider.save!
+    provider.update_authentication(:bearer => {:auth_key => options[:auth_key], :save => true})
+    valid_provider = provider.authentication_check.first
+    if valid_provider
+      self.deployed_ems = provider
+      save!
+    else
+      provider.destroy
+    end
+    valid_provider
+  end
+
+  def generate_ansible_inventory_for_subscription
+    template = <<eos
+[OSEv3:children]
+nodes
+#{"masters" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+
+[OSEv3:vars]
+ansible_ssh_user=#{ssh_auth.userid}
+deployment_type=openshift-enterprise
+rhsub_user=#{rhsm_auth.userid}
+rhsub_pass=#{rhsm_auth.password_encrypted}
+rhsub_pool=#{rhsm_auth.rhsm_sku}
+ansible_sudo=true
+
+[nodes]
+#{(roles_addresses("node") - [roles_addresses("deployment_master")]).join("\n") unless roles_addresses("node").empty?}
+
+#{"[masters]" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+#{(roles_addresses("master") - [roles_addresses("deployment_master")]).join("\n") unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+eos
+    template
+  end
+
+  def generate_roles
+    result = {}
+    roles = []
+    container_deployment_nodes.each do |node|
+      roles << node.roles
+    end
+    roles = roles.flatten.uniq
+    result["master"] = {"osm_use_cockpit" => "false"} if roles.include? "master"
+    result["node"] = {} if roles.include? "node"
+    result["storage"] = {} if roles.include? "storage"
+    result
+  end
+
+  def generate_ansible_yaml
+    hash = {}
+    hash["ansible_config"] = ANSIBLE_CONFIG_LOCATION
+    hash["ansible_log_path"] = ANSIBLE_CONFIG_LOG
+    hash["ansible_inventory_path"] = ANSIBLE_CONFIG_INVENTORY_PATH
+    hash["ansible_ssh_user"] = ssh_auth.userid
+    hash["deployment"] = {}
+    hash["deployment"]["hosts"] = container_deployment_nodes.collect(&:to_ansible_config_format)
+    hash["deployment"]["roles"] = generate_roles
+    hash["version"] = version
+    hash["variant_version"] = kind.include?("origin") ? '1.2' : '3.2'
+    hash["variant"] = kind
+    # hash[:deployment][:roles][:master][:openshift_master_identity_providers] = []
+    # hash[:deployment][:roles][:master][:openshift_master_identity_providers] << identity_provider_auth.first.send(:to_ansible_config_format)
+    unless identity_provider_auth.first.htpassd_users.empty?
+      hash[:deployment][:roles][:master][:openshift_master_htpasswd_users] = {}
+      hash[:deployment][:roles][:master][:openshift_master_htpasswd_users] = identity_provider_auth.first.htpassd_users
```

Comment:
I find it strange that you are mixing strings and symbols...is that intentional?

Also, can you remove the commented lines...doesn't seem they are needed.